### ISSUE TITLE
Inference (forward pass) with Tensorflow and OpenCV::dnn  yields very different results

### ISSUE THREAD

**Issue by gorbad-6** [opencv/opencv#10233]
I have trained an U-Net model with Keras(tf backend), and frozen/transformed the graph with the proper tensorflow tools. I can load my network into the opencv:dnn module, and make a successful forward pass. The result that I get, however, is so different from the (proper) tensorflow pass, that the opencv result becomes unusable.

##### System information (version)

- OpenCV => 3.3.1-dev
- Operating System / Platform => Ubuntu 16.04
- Compiler => gcc 5.4.0 (C++11)

##### Detailed description

The model I trained is an U-Net, built in keras, trained as NHWC, which, has loaded into opencv the following layers:
```
batch_normalization_1/batchnorm/mul_1
conv2d_1/convolution
conv2d_1/Relu
conv2d_2/convolution
conv2d_2/Relu
max_pooling2d_1/MaxPool
conv2d_3/convolution
conv2d_3/Relu
conv2d_4/convolution
conv2d_4/Relu
max_pooling2d_2/MaxPool
conv2d_5/convolution
conv2d_5/Relu
conv2d_6/convolution
conv2d_6/Relu
max_pooling2d_3/MaxPool
conv2d_7/convolution
conv2d_7/Relu
conv2d_8/convolution
conv2d_8/Relu
max_pooling2d_4/MaxPool
conv2d_9/convolution
conv2d_9/Relu
conv2d_10/convolution
conv2d_10/Relu
up_size_1/ResizeNearestNeighbor
conv2d_11/convolution
conv2d_11/Relu
concatenate_1/concat
conv2d_12/convolution
conv2d_12/Relu
conv2d_13/convolution
conv2d_13/Relu
up_size_2/ResizeNearestNeighbor
conv2d_14/convolution
conv2d_14/Relu
concatenate_2/concat
conv2d_15/convolution
conv2d_15/Relu
conv2d_16/convolution
conv2d_16/Relu
up_size_3/ResizeNearestNeighbor
conv2d_17/convolution
conv2d_17/Relu
concatenate_3/concat
conv2d_18/convolution
conv2d_18/Relu
conv2d_19/convolution
conv2d_19/Relu
up_size_4/ResizeNearestNeighbor
conv2d_20/convolution
conv2d_20/Relu
concatenate_4/concat
conv2d_21/convolution
conv2d_21/Relu
conv2d_22/convolution
conv2d_22/Relu
conv2d_23/convolution
```
In short, the model makes a few down-samples with max-pooling and up-samples them with resize-nearest neighbor (I have made sure that tf and cv use the same method, aligning centers)

now when I make an inference with tensorflow I get the image:
![tf_inf](https://user-images.githubusercontent.com/33752187/33650657-35d12afc-da63-11e7-93b1-1d194f0da8b8.png)
and with opencv 
![cls0](https://user-images.githubusercontent.com/33752187/33650674-48066eb2-da63-11e7-9ec5-a8ba20ff1db4.png)

I made some tests where I cut back the network layer-by-layer to check where the networks diverge and found that none of the outputs coincide. While small errors are expected this difference is so large, that I cannot use opencv to integrate my tf model into my code. 

Sadly, I cannot share the trained network itself, but I can provide code to reproduce the steps needed to create the network architecture, and save/freeze the model.

Any thoughts/help would be appreciated!

**Comment by gorbad-6** [comment]
related issues
https://github.com/opencv/opencv/issues/10135#issuecomment-346795778
#10065

**Comment by dkurt** [comment]
> I cut back the network layer-by-layer to check where the networks diverge and found that none of the outputs coincide.

So are you sure that networks' inputs are the same? Please share a few lines code of how you prepare inputs for TensorFlow and for OpenCV.

**Comment by gorbad-6** [comment]
@dkurt Thank you for the quick reply! I might be making an error with the data layout somewhere.
My model is trained in Keras using NHWC data shape. My input images are two-channel 512x512 arrays and I load the model and pass it to the net the following way:
```python
    import tensorflow as tf
    def load_graph(frozen_graph_filename):
        with tf.gfile.GFile(frozen_graph_filename, "rb") as f:
            graph_def = tf.GraphDef()
            graph_def.ParseFromString(f.read())

        with tf.Graph().as_default() as graph:
            tf.import_graph_def(graph_def, name="my_net")
        return graph

    graph = load_graph('frozen_opt.pb')
    x = graph.get_tensor_by_name('my_net/input_1:0')
    y = graph.get_tensor_by_name('my_net/conv2d_23/BiasAdd:0')

    image = plt.imread('totoro.png')
    image = skimage.transform.resize(image,(512,512),mode='reflect').astype('float32')
    image = np.dstack([image, image]).copy(order='C')

    with tf.Session(graph=graph) as sess:
        y_out = sess.run(y, feed_dict={x: [ 255*image.reshape(512,512,2).astype('float32')]})
```

in opencv Ioad the net and make the forward pass like:
```cpp
cv::Mat image = cv::imread( testImage, CV_LOAD_IMAGE_GRAYSCALE );
cv::resize( image, image, cv::Size( 512, 512 ) );
image.convertTo( image, CV_32F );
int blobShape[] = { 1, 2, 512, 512 };
cv::Mat inBlob( 4, blobShape, CV_32F );
image.copyTo( Mat( image.rows, image.cols, CV_32F, inBlob.ptr<float>( 0, 0 ) ) );
image.copyTo( Mat( image.rows, image.cols, CV_32F, inBlob.ptr<float>( 0, 1 ) ) );
net.setInput(inBlob,"input_1");
cv::Mat result = net.forward();
cv::Mat pred = result.reshape( 1, 1 );
float * p = pred.ptr<float>( 0 );
std::vector<float> test;
int outS = 512;
for( int k = 0; k < outS * outS; ++k )
{
        float value = *( p + k + outS * outS * ferl );
        test.push_back( value );
}
cv::Mat out_class_0 = cv::Mat( outS, outS, CV_32F, &test[0] );

```

**Comment by dkurt** [comment]
@gorbad-6, please check several things:
* Equality of `cv::resize` and `skimage.transform.resize`. Try to use the simplest interpolation strategies to achieve similar results.
* Values range. Does `plt.imread` loads images in `[0, 1]` or `[0, 255]` range? You have an additional `x255` scale at TensorFlow code.
* Output interpretation. If it's an image it still has HCWH data layout.

**Comment by gorbad-6** [comment]
@dkurt thanks for the comments,
Before submitting this issue, I checked the top two of the points you mentioned ( the resize is the same as
```python
with tf.Session() as sdd:
    tmp = image.reshape(1,image.shape[0], image.shape[1],1)
    tmp = tf.image.resize_nearest_neighbor(tmp,[512, 512], align_corners=False)
    image = tmp.eval().reshape(512,512)
```
 as well as making sure that TF uses the same resize NN method as opencv (align centers) during training.

When using the tf output from python, i take care to compare both channels of the output with both channels of the output from opencv.
```python
out_image = y_out[ 0, :, :,0] # y_out[ 0, :, :,1]
```
the output of the network is in the NHWC format when using python/tensorflow

I also played around with the interpretation of the opencv output blob. The output you see in the initial comment, was the first 512x512 floats reshaped into an image (NCHW).

 I also assumed that the result is stored in the format HWC per pixel format :
  (0,0,0), (0,0,1), (1,0,0), (1,0,1),  (row_i, col_j, channel 0), (row_i, col_j, channel 1), ...
to no avail: when i assemble the image with
```cpp
int outS = 512;
for( int k = 0; k < outS * outS; ++k )
{
        float value = *( p + k * 2 );
        test.push_back( value );
}
cv::Mat out_class_0 = cv::Mat( outS, outS, CV_32F, &test[0] );
```
I get 
![cls0](https://user-images.githubusercontent.com/33752187/33664437-eefbfaf4-da93-11e7-8fa5-4645923fc0d9.png)

**Comment by gorbad-6** [comment]
For the sake of completeness, I paste the dump from my Keras model as well
____________________________________________________________________________________________________
Layer (type)                     Output Shape          Param #     Connected to                     
====================================================================================================
input_1 (InputLayer)             (None, 512, 512, 2)   0                                            
____________________________________________________________________________________________________
batch_normalization_1 (BatchNorm (None, 512, 512, 2)   8           input_1[0][0]                    
____________________________________________________________________________________________________
conv2d_1 (Conv2D)                (None, 512, 512, 64)  1216        batch_normalization_1[0][0]      
____________________________________________________________________________________________________
conv2d_2 (Conv2D)                (None, 512, 512, 64)  36928       conv2d_1[0][0]                   
____________________________________________________________________________________________________
dropout_1 (Dropout)              (None, 512, 512, 64)  0           conv2d_2[0][0]                   
____________________________________________________________________________________________________
max_pooling2d_1 (MaxPooling2D)   (None, 256, 256, 64)  0           dropout_1[0][0]                  
____________________________________________________________________________________________________
conv2d_3 (Conv2D)                (None, 256, 256, 128) 73856       max_pooling2d_1[0][0]            
____________________________________________________________________________________________________
conv2d_4 (Conv2D)                (None, 256, 256, 128) 147584      conv2d_3[0][0]                   
____________________________________________________________________________________________________
dropout_2 (Dropout)              (None, 256, 256, 128) 0           conv2d_4[0][0]                   
____________________________________________________________________________________________________
max_pooling2d_2 (MaxPooling2D)   (None, 128, 128, 128) 0           dropout_2[0][0]                  
____________________________________________________________________________________________________
conv2d_5 (Conv2D)                (None, 128, 128, 256) 295168      max_pooling2d_2[0][0]            
____________________________________________________________________________________________________
conv2d_6 (Conv2D)                (None, 128, 128, 256) 590080      conv2d_5[0][0]                   
____________________________________________________________________________________________________
dropout_3 (Dropout)              (None, 128, 128, 256) 0           conv2d_6[0][0]                   
____________________________________________________________________________________________________
max_pooling2d_3 (MaxPooling2D)   (None, 64, 64, 256)   0           dropout_3[0][0]                  
____________________________________________________________________________________________________
conv2d_7 (Conv2D)                (None, 64, 64, 512)   1180160     max_pooling2d_3[0][0]            
____________________________________________________________________________________________________
conv2d_8 (Conv2D)                (None, 64, 64, 512)   2359808     conv2d_7[0][0]                   
____________________________________________________________________________________________________
dropout_4 (Dropout)              (None, 64, 64, 512)   0           conv2d_8[0][0]                   
____________________________________________________________________________________________________
max_pooling2d_4 (MaxPooling2D)   (None, 32, 32, 512)   0           dropout_4[0][0]                  
____________________________________________________________________________________________________
conv2d_9 (Conv2D)                (None, 32, 32, 1024)  4719616     max_pooling2d_4[0][0]            
____________________________________________________________________________________________________
conv2d_10 (Conv2D)               (None, 32, 32, 1024)  9438208     conv2d_9[0][0]                   
____________________________________________________________________________________________________
dropout_5 (Dropout)              (None, 32, 32, 1024)  0           conv2d_10[0][0]                  
____________________________________________________________________________________________________
up_size_1 (UpSize)               (None, 64, 64, 1024)  0           dropout_5[0][0]                  
____________________________________________________________________________________________________
conv2d_11 (Conv2D)               (None, 64, 64, 512)   4719104     up_size_1[0][0]                  
____________________________________________________________________________________________________
concatenate_1 (Concatenate)      (None, 64, 64, 1024)  0           dropout_4[0][0]                  
                                                                   conv2d_11[0][0]                  
____________________________________________________________________________________________________
conv2d_12 (Conv2D)               (None, 64, 64, 512)   4719104     concatenate_1[0][0]              
____________________________________________________________________________________________________
conv2d_13 (Conv2D)               (None, 64, 64, 512)   2359808     conv2d_12[0][0]                  
____________________________________________________________________________________________________
dropout_6 (Dropout)              (None, 64, 64, 512)   0           conv2d_13[0][0]                  
____________________________________________________________________________________________________
up_size_2 (UpSize)               (None, 128, 128, 512) 0           dropout_6[0][0]                  
____________________________________________________________________________________________________
conv2d_14 (Conv2D)               (None, 128, 128, 256) 1179904     up_size_2[0][0]                  
____________________________________________________________________________________________________
concatenate_2 (Concatenate)      (None, 128, 128, 512) 0           dropout_3[0][0]                  
                                                                   conv2d_14[0][0]                  
____________________________________________________________________________________________________
conv2d_15 (Conv2D)               (None, 128, 128, 256) 1179904     concatenate_2[0][0]              
____________________________________________________________________________________________________
conv2d_16 (Conv2D)               (None, 128, 128, 256) 590080      conv2d_15[0][0]                  
____________________________________________________________________________________________________
dropout_7 (Dropout)              (None, 128, 128, 256) 0           conv2d_16[0][0]                  
____________________________________________________________________________________________________
up_size_3 (UpSize)               (None, 256, 256, 256) 0           dropout_7[0][0]                  
____________________________________________________________________________________________________
conv2d_17 (Conv2D)               (None, 256, 256, 128) 295040      up_size_3[0][0]                  
____________________________________________________________________________________________________
concatenate_3 (Concatenate)      (None, 256, 256, 256) 0           dropout_2[0][0]                  
                                                                   conv2d_17[0][0]                  
____________________________________________________________________________________________________
conv2d_18 (Conv2D)               (None, 256, 256, 128) 295040      concatenate_3[0][0]              
____________________________________________________________________________________________________
conv2d_19 (Conv2D)               (None, 256, 256, 128) 147584      conv2d_18[0][0]                  
____________________________________________________________________________________________________
dropout_8 (Dropout)              (None, 256, 256, 128) 0           conv2d_19[0][0]                  
____________________________________________________________________________________________________
up_size_4 (UpSize)               (None, 512, 512, 128) 0           dropout_8[0][0]                  
____________________________________________________________________________________________________
conv2d_20 (Conv2D)               (None, 512, 512, 64)  73792       up_size_4[0][0]                  
____________________________________________________________________________________________________
concatenate_4 (Concatenate)      (None, 512, 512, 128) 0           dropout_1[0][0]                  
                                                                   conv2d_20[0][0]                  
____________________________________________________________________________________________________
conv2d_21 (Conv2D)               (None, 512, 512, 64)  73792       concatenate_4[0][0]              
____________________________________________________________________________________________________
conv2d_22 (Conv2D)               (None, 512, 512, 64)  36928       conv2d_21[0][0]                  
____________________________________________________________________________________________________
conv2d_23 (Conv2D)               (None, 512, 512, 2)   130         conv2d_22[0][0]                  
====================================================================================================
Total params: 34,512,842
Trainable params: 34,512,838
Non-trainable params: 4

**Comment by dkurt** [comment]
@gorbad-6, May I ask you try to run OpenCV from python? This way we'll check that inputs are truly the same. You said that an every layer produces different results, right? So the problem might be with the inputs, at the first layer (batch norm) or with outputs interpretation. Let's run both models in the same script up to first batch normalization layer and compare outputs.

**Comment by gorbad-6** [comment]
@dkurt, Thanks, your comment put me on the right trail! I usually pull and build opencv 1/month, so I thought everything is fine, but then I saw that there is a diff in the BatchNormLayer that I do not have. After a pull & build, now TF and CV give (almost) the same result.
For the sake of completeness, I paste here my testing code, in case someone has similar issues.
```python
import numpy as np
import tensorflow as tf
import cv2
import platform
import matplotlib.pylab as plt

def load_graph(frozen_graph_filename):
    with tf.gfile.GFile(frozen_graph_filename, "rb") as f:
        graph_def = tf.GraphDef()
        graph_def.ParseFromString(f.read())
    with tf.Graph().as_default() as graph:
        tf.import_graph_def(graph_def, name="my_net")
    return graph
def rgb2gray(rgb):
    return np.dot(rgb[...,:3], [0.299, 0.587, 0.114])


graph = load_graph('frozen_opt.pb')
x = graph.get_tensor_by_name('my_net/input_1:0')
y = graph.get_tensor_by_name('my_net/conv2d_23/BiasAdd:0')

image = rgb2gray(plt.imread('deployment/index.png'))
with tf.Session() as sdd:
    tmp = image.reshape(1,image.shape[0], image.shape[1],1)
    tmp = tf.image.resize_nearest_neighbor(tmp,[512, 512], align_corners=False)
    image = tmp.eval().reshape(512,512)
image = np.dstack([image, image]).copy(order='C')

y_out = None
with tf.Session(graph=graph) as sess:
        y_out = sess.run(y, feed_dict={x: [255 * image.reshape(512,512,2).astype('float32')]})

net = cv2.dnn.readNetFromTensorflow('frozen_opt.pb')
tmp = 255 * image.reshape(1, 512, 512, 2).astype('float32')
tmp = tmp.transpose((0, 3, 1, 2))

net.setInput(tmp,'input_1')
res = net.forward()

print("TF: ",tf.__version__,"CV: ", cv2.__version__,"PY: ", platform.python_version())
print((res.transpose((0, 2, 3, 1)) - y_out).flatten().sum() / (512 * 512))

>>> TF:  1.4.0 CV:  3.3.1-dev PY:  3.5.2
>>> 4.44125862487e-06
```
case closed.
